opi/__init__.py

```diff
@@ -101,18 +105,41 @@ def add_repo(filename, name, url, enabled=True, gpgcheck=True, gpgkey=None, repo
 	subprocess.call(['sudo', 'cp', tf.name, '/etc/zypp/repos.d/%s.repo' % filename])
 	subprocess.call(['sudo', 'chmod', '644', '/etc/zypp/repos.d/%s.repo' % filename])
 	tf.file.close()
-	refresh_cmd = ['sudo', 'zypper']
-	if auto_import_key:
-		refresh_cmd.append('--gpg-auto-import-keys')
-	refresh_cmd.append('ref')
+	refresh_cmd = []
+	if get_backend() == BackendConstants.zypp:
+		refresh_cmd = ['sudo', 'zypper']
+		if auto_import_key:
+			refresh_cmd.append('--gpg-auto-import-keys')
+		refresh_cmd.append('ref')
+	elif get_backend() == BackendConstants.dnf:
+		refresh_cmd = ['sudo', 'dnf', 'ref']
 	subprocess.call(refresh_cmd)
 
-def install_packages(packages, from_repo=None, flags=None):
-	args = ['sudo', 'zypper', 'in']
-	if from_repo:
-		args.extend(['--from', from_repo])
+def install_packages(packages, from_repo=None, flags=None, allow_vendor_change=False, allow_arch_change=False, allow_downgrade=False, allow_name_change=False):
+	if get_backend() == BackendConstants.zypp:
+		args = ['sudo', 'zypper', 'in']
+		if from_repo:
+			args.extend(['--from', from_repo])
+	elif get_backend() == BackendConstants.dnf:
+		args = ['sudo', 'dnf', 'in']
+		if from_repo:
+			args.extend(['--repo', from_repo])
 	if flags:
 		args.extend(flags)
+	if get_backend() == BackendConstants.zypp:
+		if allow_downgrade:
+			args.extend(['--allow-downgrade'])
```

Comment:
If you only add a single element, better use `append`:

```suggestion
			args.append('--allow-downgrade')
```